## Creating Taxonomy

In [2]:
import pandas as pd
from collections import Counter
import spacy
from spacy.matcher import Matcher
from tqdm import tqdm
tqdm.pandas()


# Load the English model
nlp = spacy.load("en_core_web_lg")

In [5]:
df = pd.read_parquet('../../data/processed/targets/avocado_train_targets_exploded.parquet', engine='fastparquet')

In [10]:
df.shape

(1205714, 19)

In [12]:
# Filter empty entries

df_filtered = df[~df['action_object_pairs'].apply(lambda x: x is None or (isinstance(x, list) and len(x) == 0))]

In [13]:
df_filtered.shape

(722472, 19)

### Inspect Action-Object-Pairs results

In [14]:
all_words = [word for sublist in df_filtered['action_object_pairs'] for word in sublist]

# Use Counter to count occurrences of each word
word_counts = Counter(all_words)

# Convert to DataFrame (optional, if you want to keep it in tabular form)
distinct_words_df = pd.DataFrame(word_counts.items(), columns=['Word', 'Count'])
distinct_words_df = distinct_words_df.sort_values(by='Count', ascending=False)

In [17]:
distinct_words_df.head(50)

Word  Count
20           start_Server  12831
63                send_it  12052
177        have_questions  11024
64           send_message  10704
301            send_email   8186
19           fail_Message   8065
18         start_Failures   6983
376     start_Occurrences   6428
395               call_me   6177
103             thank_you   4472
234            contact_me   3995
75           post_message   3790
165         miss_UNIVERSE   3742
410               do_what   3584
675             send_mail   3490
425      start_occurrence   3152
152                 do_it   2816
943             give_call   2642
860             need_help   2541
1371   unsubscribe_mailto   2468
66                join_us   2444
8                 do_this   2408
1233        send_courtesy   2394
534              call_you   2392
94           have_meeting   2311
40              take_look   2148
485               tell_me   2049
1903               send_e   2031
798   provide_information   2006
147         notify_sender   1923
436         receive_email   1857
548             have_time   1663
424           have_access   1617
101               do_that   1549
175          make_changes   1489
223            contact_us   1457
79               help_you   1453
85              send_this   1426
145   contain_information   1423
1146        have_problems   1417
470           provide_you   1415
272      provide_services   1380
1475               use_it   1377
1618            receive_e   1338
1799           set_DIGEST   1333
869             take_care   1323
862               help_me   1322
741      need_information   1282
655            make_sense   1271
146            receive_it   1270

| Action-Object-Pair    | Count |
|-------------|-------------|
 start_Server |	12831
send_it |	12052
have_questions |	11024
send_message |	10704
send_email |	8186
fail_Message |	8065
start_Failures |	6983
start_Occurrences |	6428
call_me	| 6177
thank_you | 4472
contact_me | 3995
post_message | 3790
miss_UNIVERSE | 3742
do_what | 3584
send_mail | 3490
start_occurrence | 3152
do_it | 2816
give_call | 2642
need_help | 2541
unsubscribe_mailto | 2468

After inspecting the most-frequently occurring action-object-pairs, the following ones will not be considered further:

| Action-Object-Pair    | Reason |
|-------------|-------------|
|fail_Message|belongs to error message and thus does not contain a humanly intent|
|start_Failures|belongs to error message and thus does not contain a humanly intent|
|start_Occurrences|belongs to error message and thus does not contain a humanly intent|
|post_message|belongs to an automated message|
|unsubscribe_mailto|belongs to an automated message|
|miss_UNIVERSE|belongs to an automated message|
|start_occurrence|belongs to an automated message|

also the extremeprogramming unsubscribe message is being filtered from the original dataframe, as well as java messages